kuka_external_control_sdk/kss/include/kuka/external-control-sdk/kss/message_builder.h

```diff
@@ -0,0 +1,110 @@
+
+namespace kuka::external::control::kss {
+
+class MotionState : public BaseMotionState {
+ public:
+  MotionState(std::size_t dof) : BaseMotionState(dof) {
+    measured_positions_.resize(dof, std::numeric_limits<double>::quiet_NaN());
+    measured_torques_.resize(dof, std::numeric_limits<double>::quiet_NaN());
+    measured_velocities_.resize(dof, std::numeric_limits<double>::quiet_NaN());
+    measured_cartesian_positions_.resize(6, std::numeric_limits<double>::quiet_NaN());
+
+    first_cartesian_position_index_ += kMessagePrefix.length();
+    first_cartesian_position_index_ += kCartesianPositionsPrefix.length() - 1;
+  }
+
+  MotionState(const MotionState& other) = default;
+  MotionState& operator=(const MotionState& other);
+
+  void CreateFromXML(const char* incoming_xml);
+  int GetIpoc() { return ipoc_; }
+  int GetDelay() { return delay_; }
+
+ private:
+  const std::string kMessagePrefix = "<Rob Type=\"KUKA\">";
+
+  const std::string kCartesianPositionsPrefix = "<RIst";
+  const std::vector<std::string> kCartesianPositionAttributePrefixes = {" X=\"", " Y=\"", " Z=\"",
+                                                                        " A=\"", " B=\"", " C=\""};
+  const std::string kAttributeSuffix = "\"/>";
+
+  const std::string kJointPositionsPrefix = "<AIPos";
+
+  const std::string kDelayNodePrefix = "<Delay D=\"";
+  const std::string kIpocNodePrefix = "<IPOC>";
+  const std::string kIpocNodeSuffix = "</IPOC>";
+  const std::string kMessageSuffix = "</Rob>";
+
+  int first_cartesian_position_index_ = 0;
+
+  long ipoc_ = 0;
+  long delay_ = 0;
+
+  static constexpr int kPrecision = 6;
+};
+
+class ControlSignal : public BaseControlSignal {
+ public:
+  ControlSignal(std::size_t dof) : BaseControlSignal(dof), initial_positions_(dof) {
+    joint_position_values_.resize(dof, 0.0);
+    cartesian_position_values_.resize(6, 0.0);
+    for (int i = 1; i <= dof; ++i) {
+      joint_position_attribute_prefixes_.push_back(" A" + std::to_string(i) + "=\"");
+    }
+  }
+
+  // Create XML containing relative positions in rad
+  std::optional<std::string_view> CreateXMLString(int last_ipoc, bool stop_control = false);
+
+  void SetInitialPositions(const MotionState& initial_positions);
+  bool InitialPositionsSet() const { return initial_positions_set_; }
+
+ private:
+  void AppendToXMLString(std::string_view str);
+
+  const std::string kMessagePrefix = "<Sen Type=\"KROSHU\">";
+  const std::string kJointPositionsPrefix = "<AK";
+  std::vector<std::string> joint_position_attribute_prefixes_;
+  const std::string kJointPositionAttributeFormat = "%." + std::to_string(kPrecision) + "f";
+  const std::string kJointPositionsSuffix = "/>";
+  const std::string kStopNodePrefix = "<Stop>";
+  const std::string kStopNodeSuffix = "</Stop>";
+  const std::string kIpocNodePrefix = "<IPOC>";
+  const std::string kIpocNodeSuffix = "</IPOC>";
+  const std::string kMessageSuffix = "</Sen>";
+
+  bool initial_positions_set_ = false;

```

Comment:
I might have missed it, but is it not necessary to reset this variable if the controlling has ended? As far as I could tell, this is reset if the Robot class is reset, but it should be possible to start control multiple times with the same robot instance